File: P1/ejercicio4/PasswordSpecialCharacterFilter.java
package ejercicio4;

public class PasswordSpecialCharacterFilter implements Filter {
    // Lista privada de caracteres especiales permitidos
    private static final String SPECIAL_CHARACTERS = "!@#$%^&*(),.?\":{}|<>";

    @Override
    public void execute(Message message) throws Exception {
        String password = message.getPassword();
        if (!password.matches(".*[" + SPECIAL_CHARACTERS + "].*")) {
            throw new Exception("Contraseña inválida: debe contener al menos un carácter especial");
        }
    }
}

File: P2/email_filter_app/lib/filters/password_special_character_filter.dart
import '../credentials.dart';
import 'filter.dart';

class PasswordSpecialCharacterFilter extends Filter {
  // Lista de caracteres especiales permitidos
  static const String specialCharacters = "!@#\$%^&*(),.?\":{}|<>";

  PasswordSpecialCharacterFilter()
    : super("Validar caracteres especiales en contraseña");

  @override
  void execute(Credentials credentials) {
    final password = credentials.password;
    final regex = RegExp('.*[${RegExp.escape(specialCharacters)}].*');
    if (!regex.hasMatch(password)) {
      throw Exception(
        'Contraseña inválida: debe contener al menos un carácter especial',
      );
    }
  }
}
